tss-esapi/src/abstraction/transient.rs

```diff
@@ -172,173 +192,41 @@ impl TransientKeyContext {
-    ///
-    /// The method takes the key as a parameter and returns its public part.
-    ///
-    /// # Errors
-    /// * errors are returned if any method calls return an error: `Context::context_load`,
-    /// `Context::read_public`, `Context::flush_context`,
-    /// `TransientKeyContext::set_session_attrs`
-    pub fn read_public_key(&mut self, key_context: TpmsContext) -> Result<PublicKey> {
-        self.set_session_attrs()?;
-        let key_handle = self.context.context_load(key_context)?;
-
-        self.set_session_attrs()?;
-        let public = self
-            .context
-            .read_public(key_handle.into())
-            .or_else(|e| {
-                self.context.flush_context(key_handle)?;
-                Err(e)
-            })?
-            .0;
-
-        let key = match public {
-            Public::Rsa { unique, .. } => PublicKey::Rsa(unique.value().to_vec()),
-            Public::Ecc { unique, .. } => PublicKey::Ecc {
-                x: unique.x().value().to_vec(),
-                y: unique.y().value().to_vec(),
-            },
-            _ => return Err(Error::local_error(ErrorKind::UnsupportedParam)),
-        };
-        self.context.flush_context(key_handle)?;
-
-        Ok(key)
+        Ok(KeyMaterial {
+            public: PublicKey::Rsa(public_key.to_vec()),
+            private: vec![],
+        })
     }
 
     /// Encrypt a message with an existing key.
     ///
     /// Takes the key as a parameter, encrypts the message and returns the ciphertext. A label (i.e.
     /// nonce) can also be provided.
-    ///
-    /// # Errors
-    /// * errors are returned if any method calls return an error: `Context::context_load`,
-    /// `Context::rsa_encrypt`, `Context::flush_context`, `TransientKeyContext::set_session_attrs`
-    /// `Context::set_handle_auth`
     pub fn rsa_encrypt(
         &mut self,
-        key_context: TpmsContext,
+        key_material: KeyMaterial,
+        key_params: KeyParams,
         key_auth: Option<Auth>,
         message: PublicKeyRsa,
-        scheme: RsaDecryptionScheme,
         label: Option<Data>,
     ) -> Result<PublicKeyRsa> {
-        self.set_session_attrs()?;
-        let key_handle = self
-            .context
-            .context_load(key_context)
-            .map(KeyHandle::from)?;
-        if let Some(key_auth_value) = key_auth {
-            self.context
-                .tr_set_auth(key_handle.into(), &key_auth_value)
-                .or_else(|e| {
-                    self.context.flush_context(key_handle.into())?;
-                    Err(e)
-                })?;
-        }
-        self.set_session_attrs()?;
+        let key_handle = self.load_key(key_params, key_material, key_auth)?;
+        let decrypt_scheme = if let KeyParams::Rsa { scheme, .. } = key_params {
+            scheme.try_into()?
+        } else {
+            return Err(Error::local_error(ErrorKind::InvalidParam));
+        };
 
+        self.set_session_attrs()?;
```

Comment:
Just for my education, when and why does `self.set_session_attrs()` need to be called? The reason I ask is that this change seems to have deleted a lot of the calls, but retained a few. For example, we no longer call it immediately before `self.load_key()`, but we _do_ still call it before `self.context.rsa_encrypt()` for example. Do we only need to call it before using `self.context`, and otherwise not?

Reply:
`self.set_session_attrs()` is called to enable some flags on the sessions that we use for almost all TPM operations (so yes, before calling methods on `self.context`). We use them to enable parameter encryption over the wire between Parsec and the TPM.